### PR TITLE
Create frontend compiler in `flutter test` lazily.

### PR DESCRIPTION
**Pull request by aam**
This is needed to avoid unwarranted runtime dependencies on Artifacts, for example when in non-preview-dart-2 mode. 
Refactor the compilation-related code into _Compiler class.

### CODE REVIEW

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 1]
```diff
@@ -87,26 +87,23 @@ enum _InitialResult { crashed, timedOut, connected }
...
-  }) : assert(shellPath != null) {
+// This class is a wrapper around compiler that allows multiple isolates enqueue
+// compilation requests, but ensures only one compilation at a time.
+class _Compiler {
+  final StreamController<_CompilationRequest> compilerController;
+  ResidentCompiler compiler;
+  final List<_CompilationRequest> compilationQueue;
+
+  _Compiler() :
+        compilerController = new StreamController<_CompilationRequest>(),
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Just initialize these where you declare them?

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 2]
```diff
@@ -87,26 +87,23 @@ enum _InitialResult { crashed, timedOut, connected }
...
-    this.previewDart2,
-    this.port,
-    this.dillFilePath,
-  }) : assert(shellPath != null) {
+// This class is a wrapper around compiler that allows multiple isolates enqueue
+// compilation requests, but ensures only one compilation at a time.
+class _Compiler {
+  final StreamController<_CompilationRequest> compilerController;
+  ResidentCompiler compiler;
+  final List<_CompilationRequest> compilationQueue;
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Nit: group the `final` members together.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 3]
```diff
@@ -87,26 +87,23 @@ enum _InitialResult { crashed, timedOut, connected }
...
+
+  _Compiler() :
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Put the constructor first

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#constructors-come-first-in-a-class

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 4]
```diff
@@ -141,7 +138,33 @@ class _FlutterPlatform extends PlatformPlugin {
     }, onDone: () {
       outputDillDirectory.delete(recursive: true);
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: `deleteSync()`

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 5]
```diff
@@ -115,19 +112,19 @@ class _FlutterPlatform extends PlatformPlugin {
...
       if (isEmpty) {
         while (compilationQueue.isNotEmpty) {
-          final CompilationRequest request = compilationQueue.first;
+          final _CompilationRequest request = compilationQueue.first;
           printTrace('Compiling ${request.path}');
           final String outputPath = await compiler.recompile(request.path,
-            <String>[request.path],
-            outputPath: outputDill.path
+              <String>[request.path],
+              outputPath: outputDill.path
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Nit: trailing comma (then you can go back to 2-space indent)

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 6]
```diff
@@ -87,26 +87,23 @@ enum _InitialResult { crashed, timedOut, connected }
 enum _TestResult { crashed, harnessBailed, testBailed }
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Maybe change `dillFilePath` above (and everywhere in this file) to be `precompiledDillPath`?

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 7]
```diff
@@ -141,7 +138,33 @@ class _FlutterPlatform extends PlatformPlugin {
...
       outputDillDirectory.delete(recursive: true);
     });
+
+    compiler = new ResidentCompiler(
+        artifacts.getArtifactPath(Artifact.flutterPatchedSdkPath),
+        packagesPath: PackageMap.globalPackagesPath);
+  }
+
+
+  Future<String> compile(String mainDart) async {
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Not `async`

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 6 above)
Comment (aam) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Done

**Review by tvolkert — APPROVED** [review]
LGTM

**Pull request merged**